File: src/main/java/brightspark/runicmagic/spell/Spell.java
package brightspark.runicmagic.spell;

import brightspark.runicmagic.RunicMagic;
import brightspark.runicmagic.enums.RuneType;
import brightspark.runicmagic.enums.SpellType;
import brightspark.runicmagic.util.RunicMagicException;
import brightspark.runicmagic.util.SpellCastData;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.Vec3d;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.util.text.TextComponentTranslation;
import net.minecraft.world.World;
import net.minecraftforge.oredict.OreDictionary;
import net.minecraftforge.registries.IForgeRegistryEntry;

import java.util.HashMap;
import java.util.Map;
import java.util.Random;
import java.util.function.Predicate;

public abstract class Spell extends IForgeRegistryEntry.Impl<Spell>
{
	private final String unlocName;
	private final ResourceLocation iconRL;
	private final SpellType spellType;
	private final int level;
	//If true, then is a selectable spell for use with a staff
	//If false, then is an instant click spell (e.g. teleport spell)
	protected boolean selectable = true;
	protected int cooldown = 0;
	protected int castTime = 0;

	private Map<RuneType, Short> cost = new HashMap<>();

	public Spell(String name, SpellType spellType, int level)
	{
		setRegistryName(name);
		iconRL = new ResourceLocation(RunicMagic.MOD_ID, "textures/spells/" + name + ".png");
		this.spellType = spellType;
		this.level = level;
		unlocName = RunicMagic.MOD_ID + ".spell." + name + ".name";
	}

	/**
	 * Gets the ResourceLocation of this spell's icon
	 * @return Icon ResourceLocation
	 */
	public ResourceLocation getIconRL()
	{
		return iconRL;
	}

	/**
	 * Gets the spell type of this spell
	 * @return The spell type
	 */
	public SpellType getSpellType()
	{
		return spellType;
	}

	/**
	 * The magic level required to use this spell
	 * @return Magic level requirement
	 */
	public int getLevel()
	{
		return level;
	}

	/**
	 * Checks if the player meets the spell's requirements to start casting
	 * @param player The player
	 * @return Whether the spell can be cast
	 */
	public abstract boolean canCast(EntityPlayer player);

	/**
	 * Process updating the casting of this spell if it's not instant
	 * @param world The world
	 * @param progress The current progress in ticks since the spell was executed
	 * @return true if the spell should be cancelled
	 */
	public boolean updateCasting(World world, EntityPlayer player, int progress)
	{
		if(castTime > 0)
			throw new RunicMagicException("Non-instant casting spell %s does not override Spell#updateCasting!", this);
		return false;
	}

	/**
	 * Executes this spell and returns true if successful
	 * Will be called instantly for instant spells, or at the end of casting
	 */
	public abstract boolean execute(EntityPlayer player, SpellCastData data);

	/**
	 * Called when a spell cast is cancelled
	 */
	public void onCastCancel(EntityPlayer player)
	{
		player.sendMessage(new TextComponentString("Spell ").appendSibling(getUnlocNameTextComponent()).appendText(" cancelled!"));
	}

	/**
	 * Add a rune cost to the total cost to cast this spell
	 */
	public Spell addRuneCost(RuneType type, int amount)
	{
		cost.put(type, (short) amount);
		return this;
	}

	/**
	 * Returns a copy of the costs to cast this spell
	 */
	public Map<RuneType, Short> getCost()
	{
		return new HashMap<>(cost);
	}

	/**
	 * If this spell is selectable for use with a staff
	 */
	public boolean isSelectable()
	{
		return selectable;
	}

	/**
	 * Cooldown for this spell in ticks
	 */
	public int getCooldown()
	{
		return cooldown;
	}

	/**
	 * Time for this spell to cast in ticks
	 */
	public int getCastTime()
	{
		return castTime;
	}

	public String getUnlocName()
	{
		return unlocName;
	}

	public TextComponentTranslation getUnlocNameTextComponent()
	{
		return new TextComponentTranslation(getUnlocName());
	}

	@Override
	public String toString()
	{
		return getRegistryName().toString();
	}

	// <<<<<<<< Util methods >>>>>>>>

	protected static double getRandOffset(World world, double variance)
	{
		return variance == 0D ? 0D : (world.rand.nextDouble() * variance) - (variance / 2);
	}

	protected static Vec3d posOffset(World world, Vec3d pos, double xVary, double yVary, double zVary)
	{
		return pos.add(getRandOffset(world, xVary), getRandOffset(world, yVary), getRandOffset(world, zVary));
	}

	protected boolean hasPlayerMoved(EntityPlayer player)
	{
		return player.prevPosX != player.posX || player.prevPosY != player.posY || player.prevPosZ != player.posZ;
	}

	protected int countItems(EntityPlayer player, ItemStack stack)
	{
		int count = 0;
		for(int i = 0; i < player.inventory.getSizeInventory(); i++)
		{
			ItemStack invStack = player.inventory.getStackInSlot(i);
			if(OreDictionary.itemMatches(stack, invStack, false))
				count += invStack.getCount();
		}
		return count;
	}

	protected static Vec3d createRandVector(Random random)
	{
		double phi = random.nextDouble() * 2 * Math.PI;
		double theta = Math.acos((random.nextDouble() * 2D) - 1D);
		double x = Math.sin(theta) * Math.cos(phi);
		double y = Math.sin(theta) * Math.sin(phi);
		double z = Math.cos(theta);
		return new Vec3d(x, y, z);
	}

	protected boolean playerHasStack(EntityPlayer player, ItemStack stack, boolean strict)
	{
		return playerHasStack(player, invStack -> OreDictionary.itemMatches(stack, invStack, strict));
	}

	protected boolean playerHasStack(EntityPlayer player, Predicate<ItemStack> predicate)
	{
		for(int i = 0; i < player.inventory.getSizeInventory(); i++)
			if(predicate.test(player.inventory.getStackInSlot(i)))
				return true;
		return false;
	}

	protected boolean playerHasSpace(EntityPlayer player)
	{
		return playerHasStack(player, ItemStack::isEmpty);
	}

	protected void givePlayerStack(EntityPlayer player, ItemStack stack)
	{
		if(!player.addItemStackToInventory(stack))
			player.entityDropItem(stack, 0F);
	}
}

File: src/main/kotlin/brightspark/runicmagic/spell/Spell.kt
package brightspark.runicmagic.spell

import brightspark.runicmagic.model.RuneType
import brightspark.runicmagic.model.SpellCastData
import brightspark.runicmagic.model.SpellType
import brightspark.runicmagic.util.RunicMagicException
import brightspark.runicmagic.util.appendTranslation
import brightspark.runicmagic.util.sendMessage
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.entity.player.ServerPlayerEntity
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.vector.Vector3d
import net.minecraft.util.text.StringTextComponent
import net.minecraft.world.World
import net.minecraftforge.registries.ForgeRegistryEntry
import java.util.*
import kotlin.math.acos
import kotlin.math.cos
import kotlin.math.sin

abstract class Spell(props: Properties) : ForgeRegistryEntry<Spell>() {
	val iconRL: ResourceLocation by lazy {
		ResourceLocation(registryName!!.namespace, "textures/spells/${registryName!!.path}.png")
	}
	val unlocName: String by lazy { "${registryName!!.namespace}.spell.${registryName!!.path}.name" }

	val spellType: SpellType = props.spellType
	val level: Int = props.levelReq
	//If true, then is a selectable spell for use with a staff
	//If false, then is an instant click spell (e.g. teleport spell)
	val selectable: Boolean = props.selectable
	val cooldown: Int = props.cooldown
	val castTime: Int = props.castTime
	val runeCost: Map<RuneType, Int> = props.runeCost.toMap()

	/**
	 * Checks if the player meets the spell's requirements to start casting
	 * @param player The player
	 * @return Whether the spell can be cast
	 */
	abstract fun canCast(player: PlayerEntity): Boolean

	/**
	 * Process updating the casting of this spell if it's not instant
	 * @param world The world
	 * @param progress The current progress in ticks since the spell was executed
	 * @return true if the spell should be cancelled
	 */
	open fun updateCasting(world: World, player: PlayerEntity, progress: Int): Boolean {
		if (castTime > 0)
			throw RunicMagicException("Non-instant casting spell $this does not override Spell#updateCasting!")
		return false
	}

	/**
	 * Executes this spell and returns true if successful
	 * Will be called instantly for instant spells, or at the end of casting
	 * @return true if spell cast successfully
	 */
	abstract fun execute(player: ServerPlayerEntity, data: SpellCastData): Boolean

	/**
	 * Called when a spell cast is cancelled
	 */
	open fun onCastCancel(player: PlayerEntity) {
		// TODO: Localise
		player.sendMessage(StringTextComponent("Spell ").appendTranslation(unlocName).appendString(" cancelled!"))
	}

	override fun toString(): String = registryName.toString()

	//
	// Util methods
	//

	protected fun randOffset(world: World, variance: Double): Double =
		if (variance == 0.0) 0.0 else (world.rand.nextDouble() * variance) - (variance / 2)

	protected fun posOffset(world: World, posComponent: Double, variance: Double): Double =
		posComponent + randOffset(world, variance)

	protected fun posOffset(world: World, pos: Vector3d, xVary: Double, yVary: Double, zVary: Double): Vector3d =
		pos.add(randOffset(world, xVary), randOffset(world, yVary), randOffset(world, zVary))

	protected fun randVector(random: Random): Vector3d {
		val phi = random.nextDouble() * 2 * Math.PI
		val theta = acos((random.nextDouble() * 2.0) - 1.0)
		return Vector3d(sin(theta) * cos(phi), sin(theta) * sin(phi), cos(theta))
	}

	// TODO: Change this so only explicit player movement counts? e.g. being pushed by water or another entity doesn't count
	protected fun hasPlayerMoved(player: PlayerEntity): Boolean =
		player.prevPosX != player.posX || player.prevPosY != player.posY || player.prevPosZ != player.posZ

	protected fun countItemsInPlayerInv(player: PlayerEntity, stack: ItemStack): Int =
		(0 until player.inventory.sizeInventory).sumBy {
			val invStack = player.inventory.getStackInSlot(it)
			return@sumBy if (ItemStack.areItemsEqualIgnoreDurability(stack, invStack)) invStack.count else 0
		}

	protected fun playerHasStack(player: PlayerEntity, predicate: (ItemStack) -> Boolean): Boolean =
		(0 until player.inventory.sizeInventory).any {
			player.inventory.getStackInSlot(it).run { !isEmpty && predicate(this) }
		}

	protected fun playerHasItem(player: PlayerEntity, item: Item): Boolean = playerHasStack(player) { it.item == item }

	protected fun playerHasSpace(player: PlayerEntity): Boolean =
		(0 until player.inventory.sizeInventory).any { player.inventory.getStackInSlot(it).isEmpty }

	protected fun givePlayerStack(player: PlayerEntity, stack: ItemStack) {
		if (!player.addItemStackToInventory(stack))
			player.entityDropItem(stack)
	}

	class Properties(
		var spellType: SpellType,
		var levelReq: Int,
		var selectable: Boolean = true,
		var cooldown: Int = 0,
		var castTime: Int = 0,
		var runeCost: MutableMap<RuneType, Int> = mutableMapOf()
	) {
		fun setSpellType(spellType: SpellType): Properties {
			this.spellType = spellType
			return this
		}

		fun setLevelReq(levelReq: Int): Properties {
			this.levelReq = levelReq
			return this
		}

		fun setSelectable(selectable: Boolean): Properties {
			this.selectable = selectable
			return this
		}

		fun setCooldown(cooldown: Int): Properties {
			this.cooldown = cooldown
			return this
		}

		fun setCastTime(castTime: Int): Properties {
			this.castTime = castTime
			return this
		}

		fun addRuneCost(runeType: RuneType, amount: Int): Properties {
			runeCost[runeType] = amount
			return this
		}

		fun addRuneCost(vararg cost: Pair<RuneType, Int>): Properties {
			cost.forEach { runeCost[it.first] = it.second }
			return this
		}
	}
}
